lib/block_view.go

```diff
@@ -2919,24 +3053,72 @@ func _isEntryImmatureBlockReward(utxoEntry *UtxoEntry, blockHeight uint32, param
+	// Compute a hash of the transaction.
 	txBytes, err := txn.ToBytes(true /*preSignature*/)
 	if err != nil {
 		return errors.Wrapf(err, "_verifySignature: Problem serializing txn without signature: ")
 	}
 	txHash := Sha256DoubleHash(txBytes)
-	// Convert the txn public key into a *btcec.PublicKey
-	txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+	sig, err := btcec.ParseDERSignature(txn.Signature, btcec.S256())
+
+	// Look for the derived key in transaction ExtraData and validate it. For transactions
+	// signed using a derived key, the derived public key is passed to ExtraData.
+	var derivedPk *btcec.PublicKey
+	var derivedPkBytes []byte
+	isDerived := false
+	if txn.ExtraData != nil {
+		if derivedPkBytes, isDerived = txn.ExtraData[DerivedPublicKey]; isDerived {
+			derivedPk, err = btcec.ParsePubKey(derivedPkBytes, btcec.S256())
+			if err != nil {
+				return RuleErrorDerivedKeyInvalidExtraData
+			}
+		}
+	}
+
+	// Get the owner public key and attempt turning it into *btcec.PublicKey.
+	ownerPkBytes := txn.PublicKey
+	ownerPk, err :=  btcec.ParsePubKey(ownerPkBytes, btcec.S256())
 	if err != nil {
-		return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		return errors.Wrapf(err, "_verifySignature: Problem parsing owner public key: ")
 	}
-	// Verify that the transaction is signed by the specified key.
-	if txn.Signature == nil || !txn.Signature.Verify(txHash[:], txnPk) {
-		return RuleErrorInvalidTransactionSignature
+
+	// If no derived key is present in ExtraData, we check if transaction was signed by the owner.
+	// If derived key is present in ExtraData, we check if transaction was signed by the derived key.
+	if isDerived == false {
+		// Verify that the transaction is signed by the specified key.
+		if sig.Verify(txHash[:], ownerPk) {
+			return nil
+		}
+	} else {
+		// Look for a derived key entry in UtxoView and DB, check if it exists nor is deleted.
+		derivedKeyEntry := bav._getDerivedKeyMappingForOwner(ownerPkBytes, derivedPkBytes)
+		if derivedKeyEntry == nil || derivedKeyEntry.isDeleted {
+			return RuleErrorDerivedKeyNotAuthorized
+		}
+
+		// Return an error if transaction public keys don't match derivedKeyEntry public keys.
+		if !reflect.DeepEqual(ownerPkBytes, derivedKeyEntry.OwnerPublicKey[:]) ||
+			!reflect.DeepEqual(derivedPkBytes, derivedKeyEntry.DerivedPublicKey[:]) {
+			return RuleErrorDerivedKeyNotAuthorized
+		}
+
+		// At this point, we know the derivedKeyEntry that we have is matching.
+		// We check if the derived key hasn't been de-authorized or hasn't expired.
+		if derivedKeyEntry.OperationType != AuthorizeDerivedKeyOperationValid ||
+			derivedKeyEntry.ExpirationBlock <= uint64(blockHeight) {
+			return RuleErrorDerivedKeyNotAuthorized
+		}
+
+		// All checks passed so we try to verify the signature.
+		if sig.Verify(txHash[:], derivedPk) {
+			return nil
+		}
+
+		return RuleErrorDerivedKeyNotAuthorized
```

Comment:
Wow man! I think this works, and it's braindead simple. Great work! The hard part is over. Now onto the rest :)